src/main/java/org/ngafid/common/TimeUtils.java

```diff
@@ -259,4 +282,171 @@ public static Date addMilliseconds(Date date, Integer milliseconds) {
         cal.add(Calendar.MILLISECOND, milliseconds);
         return cal.getTime();
     }
+
+
+    public static LocalDateTimeResult calculateLocalDateTimeFromTimeSeries(
+            TimeZoneMap map,
+            StringTimeSeries utcDates,
+            StringTimeSeries utcTimes,
+            DoubleTimeSeries latitudes,
+            DoubleTimeSeries longitudes) throws Exception {
+
+        // Validate that all time series are of the same size.
+        if (utcDates.size() != utcTimes.size() || utcDates.size() != latitudes.size() || utcDates.size() != longitudes.size()) {
+            throw new IllegalArgumentException("All input time series must have the same size.");
+        }
+
+        // Prepare lists for the results.
+        ArrayList<String> localDates = new ArrayList<>(utcDates.size());
+        ArrayList<String> localTimes = new ArrayList<>(utcDates.size());
+        ArrayList<String> utcOffsets = new ArrayList<>(utcDates.size());
+
+        // Iterate over each index and calculate the corresponding local date, time, and offset.
+        for (int i = 0; i < utcDates.size(); i++) {
+            String dateTimeString = utcDates.get(i) + " " + utcTimes.get(i);
+            LocalDateTime utcDateTime = null;
+
+            // Try parsing using a formatter.
+            for (DateTimeFormatter formatter : DATE_FORMATTERS) {
+                try {
+                    utcDateTime = LocalDateTime.parse(dateTimeString, formatter);
+                    break;
+                } catch (DateTimeParseException e) {
+                    // Continue trying other formats
+                }
+            }
```

Comment:
Determine formatter once instead of doing it once for every element inside of utcDates.

Reply:
Please use this to determine the formatter outside the body of the loop.
```
DateTimeFormatter formatter = null;
var dateTimeString = utcDates.get(0) + " " + utcTimes.get(0);
for (DateTimeFormatter candidateFormatter : DATE_FORMATTERS) {
    try {
        LocalDateTime.parse(dateTimeString, candidateFormatter);

        // Previous line didnt throw an error so this is the format we want.
        formatter = candidateFormatter;
        break;
    } catch (DateTimeParseException e) {
        // Continue trying other formats
    }
}

if (formatter == null) // None of our formats worked, should throw some sort of exception.
```